### PR TITLE
refactor speedtests handling, ensure rewards only include tests up until epoch end

### PR DESCRIPTION
**Pull request by andymck**
Refactor the speedtests to prevent scenario whereby the rewards can use reports from after the epoch end.  To accommodate this I had to explode the speedtests reports in the DB from 1 row containing an array of up to 6 reports out into 1 row per report.

At the same time i have split the averaging code out from the speedtest module into its own module for better readability and maintainability.

- [x] Add migration to explode speedtests out to 1 test per db row
- [x] delete speedtests after rewarding ( ensure we keep 48hrs worth of tests )
- [x] handle todo's
- [x] test locally
- [x] tidy up
- [x] add indexes to speedtest table

### CODE REVIEW

File: mobile_verifier/migrations/15_speedtests_one_to_one.sql
Comment (andymck) [inline mobile_verifier/migrations/15_speedtests_one_to_one.sql:15]: I should change this to drop the old table but for safety sake renaming for now.  Will change to drop before merge

File: mobile_verifier/src/speedtests.rs
Comment (maplant) [inline mobile_verifier/src/speedtests.rs:195]: This function should have the same signature as `clear_heartbeats`; you should clear out the speedtests that are before the reward period

File: mobile_verifier/src/speedtests.rs
Comment (maplant) [inline mobile_verifier/src/speedtests.rs:103]: I think that using a try fold, and separating out the functions is less clear and incongruous with how heartbeats (and eventually coverage object) daemons are structured, this function should match how the heartbeats function operates. That is, have a `validate_speedtests` function on `Speedtest` that you can call `write` and `save` on in a while loop.  

Basically, I think that the structure of the function before doesn't need to be changed, because it matches the way heartbeats work and how the coverage object work. Changing this function to look differently makes the function less clear

File: mobile_verifier/src/speedtests_average.rs
Comment (maplant) [inline mobile_verifier/src/speedtests_average.rs:99]: Instead of having speedtests be an argument, it makes more sense to me to include that in `SpeedtestAverage`

File: mobile_verifier/src/speedtests_average.rs
Comment (maplant) [inline mobile_verifier/src/speedtests_average.rs:226]: This essentially force the executor to be a `PgPool`, so you might as well just make it that to clean up the argument

File: mobile_verifier/src/speedtests.rs
Comment (maplant) [inline mobile_verifier/src/speedtests.rs:129]:
```suggestion
pub async fn get_latest_speedtests_for_pubkey(
```

File: mobile_verifier/src/speedtests.rs
Comment (andymck) [reply mobile_verifier/src/speedtests.rs:204]: addressed here: a9b971c45caf62acb2c7b3c439ab824bad3ba01e

File: mobile_verifier/src/speedtests_average.rs
Comment (andymck) [reply mobile_verifier/src/speedtests_average.rs:96]: The averages are utilzed in multiple points in the code, but the only time the source speedtests from which the average was derived is required is at the point of writing out the proto.  Seems overkill to be throwing the vec of speedtests around unnecessarily

File: mobile_verifier/src/speedtests_average.rs
Comment (andymck) [reply mobile_verifier/src/speedtests_average.rs]: addressed here: a9b971c45caf62acb2c7b3c439ab824bad3ba01e
NOTE: heartbeats uses the same approach with the executor ( via HeartbeatReward::validated )

File: mobile_verifier/src/speedtests.rs
Comment (andymck) [reply mobile_verifier/src/speedtests.rs:96]: The gateway check was there prior to current changes ( https://github.com/helium/oracles/blob/main/mobile_verifier/src/speedtests.rs#L193 ).  Also the save in the changes *is* conditional, it only saves if it is validated

File: mobile_verifier/migrations/15_speedtests_one_to_one.sql
Comment (andymck) [reply mobile_verifier/migrations/15_speedtests_one_to_one.sql:17]: i am grand with that

File: mobile_verifier/src/speedtests.rs
Comment (andymck) [reply mobile_verifier/src/speedtests.rs]: i have refactored this again and kept it simple.  I was trying to shoehorn this into the prior pattern as you suggested ( and have a `validate_speedtests` iterator for example ) but it just becomes unnecessarily complicated and required boiler plate `impl` for the sake of them.  

Re adding a new proto for speedtest validity I am grand with that but suggest doing that as a separate PR.

File: mobile_verifier/src/heartbeats.rs
Comment (maplant) [inline mobile_verifier/src/heartbeats.rs:293]: I would recommend not taking a period here since it is unnecessary, have this be `reward_start: DateTime<Utc>` (same for clear_speedtests)

Comment (andymck) [reply mobile_verifier/src/heartbeats.rs]: addressed here: ae8fb0607f4b32f6881a7ce46e00551de941f410

**Review by bbalser — APPROVED** [review]

**Review by bbalser — APPROVED** [review]

File: mobile_verifier/migrations/15_speedtests_one_to_one.sql
Comment (jeffgrunewald) [inline mobile_verifier/migrations/15_speedtests_one_to_one.sql:1]:
```suggestion
```

File: mobile_verifier/migrations/15_speedtests_one_to_one.sql
Comment (jeffgrunewald) [inline mobile_verifier/migrations/15_speedtests_one_to_one.sql:7]: Any issues with this column name being a keyword in Postgres? Should it be named `serial_num` or something?

File: mobile_verifier/migrations/15_speedtests_one_to_one.sql
Comment (jeffgrunewald) [inline mobile_verifier/migrations/15_speedtests_one_to_one.sql:16]: Is the second `as tmp` necessary since you don’t refer to a `tmp` table anywhere?

File: mobile_verifier/migrations/15_speedtests_one_to_one.sql
Comment (andymck) [reply mobile_verifier/migrations/15_speedtests_one_to_one.sql:7]: Never even noticed this, but all worked fine on local testing.   Still, prob best to rename and avoid

File: mobile_verifier/src/reward_shares.rs
Comment (jeffgrunewald) [inline mobile_verifier/src/reward_shares.rs:1]:
```suggestion
```

File: mobile_verifier/src/reward_shares.rs
Comment (jeffgrunewald) [inline mobile_verifier/src/reward_shares.rs:3]:
```suggestion
    data_session::HotspotMap, heartbeats::HeartbeatReward, speedtests_average::{SpeedtestAverage, SpeedtestAverages},
```

File: mobile_verifier/src/reward_shares.rs
Comment (jeffgrunewald) [inline mobile_verifier/src/reward_shares.rs:329]:
```suggestion
```
This line should be unnecessary since you `use super::*` to bring in those imports from the parent module already

**Review by maplant — APPROVED** [review]

File: mobile_verifier/src/rewarder.rs
Comment (jeffgrunewald) [inline mobile_verifier/src/rewarder.rs:217]: Why do we clear all the other records from the end of the reward period and before, but the heartbeats we clear from the _start_ of the period and prior? That would leave the heartbeats from the each reward period in the DB for the next reward period’s evaluation each day, no?

File: mobile_verifier/src/speedtests.rs
Comment (jeffgrunewald) [inline mobile_verifier/src/speedtests.rs:27]: Not sure if file_store has sqlx already defined but you should be able to derive the FromRow against the CellSpeedtest and then your queries to grab Speedtest structs just query_as CellSpeedtest and map the results to set that result as the value of the report field in the Speedtest struct. The only real data conversion I see here is the integer types and deriving FromRow allows you to annotate integer fields to `try_from` to convert from signed to unsigned integer types

File: mobile_verifier/src/speedtests.rs
Comment (jeffgrunewald) [inline mobile_verifier/src/speedtests.rs:91]: Do you need this clone here? Looks like you only pass it by reference for the rest of the operation

File: mobile_verifier/src/speedtests.rs
Comment (jeffgrunewald) [inline mobile_verifier/src/speedtests.rs:122]: Do you need to clone the pubkeybin to insert the record? Same question with the text/string serial

File: mobile_verifier/src/speedtests.rs
Comment (jeffgrunewald) [inline mobile_verifier/src/speedtests.rs:133]: Since you’re not performing any operations on the steamed result in the `while` loop you should be able to just make this a `fetch_all` instead of a `fetch` and get the results back in a vec without the while loop

File: mobile_verifier/src/speedtests.rs
Comment (jeffgrunewald) [inline mobile_verifier/src/speedtests.rs:157]:
```suggestion
    // use latest speedtest which are no older than N hours, defined by SPEEDTEST_LAPSE
```

File: mobile_verifier/src/speedtests_average.rs
Comment (jeffgrunewald) [inline mobile_verifier/src/speedtests_average.rs:43]: Instead of just overwriting the pubkey/id each iteration can you set it once based on the first element of the loop or else pass it in as an additional argument and then filter in the iterator loop based on each speedtest having the same pubkey?

File: mobile_verifier/src/speedtests_average.rs
Comment (jeffgrunewald) [inline mobile_verifier/src/speedtests_average.rs:109]: This should probably implement the timestamp_millis feature so you don’t have to cast it to a u64, right?

File: mobile_verifier/src/speedtests_average.rs
Comment (jeffgrunewald) [inline mobile_verifier/src/speedtests_average.rs:116]: guaranteed safe unwrap?

File: mobile_verifier/src/speedtests_average.rs
Comment (jeffgrunewald) [inline mobile_verifier/src/speedtests_average.rs:130]: Move into the test sub module and then remove the dead_code annotation? Or is this used in multiple testing modules?

File: mobile_verifier/src/speedtests_average.rs
Comment (jeffgrunewald) [inline mobile_verifier/src/speedtests_average.rs:5]:
```suggestion
use helium_proto::services::poc_mobile::proto;
```
Looks like you’re (typically) referencing the SpeedtestAvgValidity as `proto::SpeedtestAvgValidity` anyway so you can clean up this import a bit

File: mobile_verifier/src/speedtests_average.rs
Comment (andymck) [reply mobile_verifier/src/speedtests_average.rs]: copy pasted from the original speedtests module but changed now to use the timestamp trait.  Returns seconds tho not millis

File: mobile_verifier/src/speedtests_average.rs
Comment (andymck) [reply mobile_verifier/src/speedtests_average.rs:119]: I believe so and copy pasted from original, so always been there.

File: mobile_verifier/migrations/15_speedtests_one_to_one.sql
Comment (andymck) [reply mobile_verifier/migrations/15_speedtests_one_to_one.sql:1]: addressed here: 67effc3558d2cd2c56ef07956dba0b19b02ffd95

File: mobile_verifier/src/rewarder.rs
Comment (andymck) [reply mobile_verifier/src/rewarder.rs:217]: Both heartbeats and speedtests were meant to use the epoch start, not end.  Speedtests clearing now fixed:
addressed here: 67effc3558d2cd2c56ef07956dba0b19b02ffd95

File: mobile_verifier/src/speedtests.rs
Comment (andymck) [reply mobile_verifier/src/speedtests.rs:133]: addressed here: 67effc3558d2cd2c56ef07956dba0b19b02ffd95

File: mobile_verifier/src/speedtests.rs
Comment (andymck) [reply mobile_verifier/src/speedtests.rs:122]: both required

File: mobile_verifier/migrations/15_speedtests_one_to_one.sql
Comment (andymck) [reply mobile_verifier/migrations/15_speedtests_one_to_one.sql]: addressed here: 67effc3558d2cd2c56ef07956dba0b19b02ffd95

File: mobile_verifier/src/speedtests.rs
Comment (andymck) [reply mobile_verifier/src/speedtests.rs]: addressed here: 67effc3558d2cd2c56ef07956dba0b19b02ffd95

File: mobile_verifier/src/speedtests_average.rs
Comment (andymck) [reply mobile_verifier/src/speedtests_average.rs]: addressed here: 67effc3558d2cd2c56ef07956dba0b19b02ffd95

File: mobile_verifier/src/speedtests_average.rs
Comment (andymck) [reply mobile_verifier/src/speedtests_average.rs]: addressed here: 67effc3558d2cd2c56ef07956dba0b19b02ffd95

File: mobile_verifier/src/speedtests_average.rs
Comment (andymck) [reply mobile_verifier/src/speedtests_average.rs]: addressed here: 67effc3558d2cd2c56ef07956dba0b19b02ffd95

File: mobile_verifier/src/reward_shares.rs
Comment (andymck) [reply mobile_verifier/src/reward_shares.rs]: addressed here: 67effc3558d2cd2c56ef07956dba0b19b02ffd95

File: mobile_verifier/src/speedtests.rs
Comment (jeffgrunewald) [reply mobile_verifier/src/speedtests.rs:91]: you can just pass by reference as `&speedtest_report.report.pubkey;`

File: mobile_verifier/src/speedtests_average.rs
Comment (jeffgrunewald) [reply mobile_verifier/src/speedtests_average.rs:46]: here's what i was thinking:
```
impl<'a, I: ?Sized> From<&'a I> for SpeedtestAverage
where
    &'a I: IntoIterator<Item = &'a Speedtest>,
{
    fn from(iter: &'a I) -> Self {
        let mut window_size = 0;
        let mut sum_upload = 0;
        let mut sum_download = 0;
        let mut sum_latency = 0;

        let mut speedtest_iter = speedtests_without_lapsed(iter.into_iter(), Duration::hours(SPEEDTEST_LAPSE)).peekable();
        let id = speedtest_iter
            .peek()
            .map(|Speedtest { report, .. }| report.pubkey.as_ref().to_vec()).unwrap_or_default();
        for Speedtest { report, .. } in speedtest_iter 
        {
            if report.pubkey.as_ref().to_vec() == id {
                sum_upload += report.upload_speed;
                sum_download += report.download_speed;
                sum_latency += report.latency;
                window_size += 1;
            }
        }

        if window_size > 0 {
            let upload_speed_avg_bps = sum_upload / window_size as u64;
            let download_speed_avg_bps = sum_download / window_size as u64;
            let latency_avg_ms = sum_latency / window_size as u32;
            let validity = validity(
                window_size as usize,
                upload_speed_avg_bps,
                download_speed_avg_bps,
                latency_avg_ms,
            );
            let tier = SpeedtestTier::new(
                window_size as usize,
                upload_speed_avg_bps,
                download_speed_avg_bps,
                latency_avg_ms,
            );
            let reward_multiplier = tier.into_multiplier();
            SpeedtestAverage {
                pubkey: id.into(),
                window_size: window_size as usize,
                upload_speed_avg_bps,
                download_speed_avg_bps,
                latency_avg_ms,
                validity,
                reward_multiplier,
            }
        } else {
            SpeedtestAverage {
                pubkey: id.into(),
                window_size: 0,
                upload_speed_avg_bps: sum_upload,
                download_speed_avg_bps: sum_download,
                latency_avg_ms: sum_latency,
                validity: proto::SpeedtestAvgValidity::TooFewSamples,
                reward_multiplier: Decimal::ZERO,
            }
        }
    }
}
```
logically this works and it compiles. i'm not a fan of the `unwrap_or_default` since that _theoretically_ could result in aggregating the speedtest average under a `SpeedtestAverage` with an empty Vec<u8> as the `Pubkey` but since that seems really unlikely that we'd have a valid `CellSpeedtest` with an empty pubkey than you should be safe to just replace this with an `unwrap()`.

i like the `peekable` implementation on the iterator to grab the first element of the list without consuming it and then checking subsequent members of the list to ensure the pubkeys match before including them in the average, but that's probably unnecessary and if overwriting the id on each iteration of the for loop over the collection of CellSpeedtests isn't an expensive operation this is probably unnecessary so i'll leave it to you to take or ignore

**Review by jeffgrunewald — APPROVED** [review]

**Pull request merged**
